include/mrdox/Metadata/Template.hpp

```diff
@@ -14,29 +14,48 @@
 
 #include <mrdox/Platform.hpp>
 #include <mrdox/Metadata/TemplateParam.hpp>
-#include <llvm/ADT/Optional.h>
+#include <mrdox/Metadata/TemplateArg.hpp>
 
 namespace clang {
 namespace mrdox {
 
-struct TemplateSpecializationInfo
+enum class TemplateSpecKind
 {
-    // Indicates the declaration that this specializes.
-    SymbolID SpecializationOf;
-
-    // Template parameters applying to the specialized record/function.
-    std::vector<TemplateParamInfo> Params;
+    Primary,
+    Explicit,
+    Partial
 };
 
-// Records the template information for a struct or function that is a template
-// or an explicit template specialization.
+// stores information pertaining to primary template and
+// partial/explicit specialization declarations
 struct TemplateInfo
 {
-    // May be empty for non-partial specializations.
+    // for primary templates:
+    //     - Params will be non-empty
+    //     - Args will be empty
+    // for explicit specializations:
+    //     - Params will be empty
+    // for partial specializations:
+    //     - Params will be non-empty
+    //     - each template parameter will appear at least
+    //       once in Args outside of a non-deduced context 
     std::vector<TemplateParamInfo> Params;
+    std::vector<TemplateArgInfo> Args;
+
+    // stores the USR of the corresponding primary template
+    // for partial and explicit specializations
+    std::optional<SymbolID> Primary;
```

Comment:
come on man this could just be `SymbolID` without the `optional` and just compare it to `EmptySID` to know if it is disengaged

Reply:
Fair

Reply:
Will change to using it once I get this refactor finished